### PR TITLE
Use blessings when possible for WPT UI

### PR DESCRIPTION
**Pull request by mrobinson**
Use blessings when it is possible for the WPT UI and fall back to using
raw control characters. This also fixes a bug where the UI didn't work
correctly in iTerm.app. Remove the one line version of the UI, since it
is no longer used as the iTerm fallback.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8551)

### CODE REVIEW

**Comment by mrobinson** [comment]
@jdm or @jgraham r?

File: tests/wpt/grouping_formatter.py [hunk 1]
```diff
@@ -17,16 +17,22 @@ def __init__(self):
...
-        # so only print one line and rely only on backspace characters.
-        self.one_line = os.environ.get("TERM_PROGRAM", "") == "iTerm.app"
+        # TODO(mrobinson, 8313): We need to add support for Windows terminals here.
+        if self.interactive:
+            try:
+                import blessings
+                self.terminal = blessings.Terminal()
+                self.clear_eol = self.terminal.clear_eol
+                self.move_up = self.terminal.move_up
+            except:
```
Comment (frewsxcv) [inline tests/wpt/grouping_formatter.py]: Can we make this `except ImportError` just so we silence+catch *all* exceptions in case one occurs from something else?

Comment (mrobinson) [inline tests/wpt/grouping_formatter.py]: The idea here was that if the import failed, if creating the Terminal failed, or if the ```__getattr__``` call in blessings fails (which <a href="https://github.com/erikrose/blessings/blob/master/blessings/__init__.py#L157">just calls into curses</a>), we should always fall back. If you like I can split the work into two try/except blocks and print an error in the second one.

Comment (jgraham) [inline tests/wpt/grouping_formatter.py]: try and except with no condition is never a good idea; even `except Exception:` is better because it won't catch `KeyboardInterrupt`. In this case I agree that splitting into two blocks is a better idea.

**Comment by jgraham** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

**Comment by mrobinson** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[tests/wpt/grouping_formatter.py, line 33 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3GRjeGJL9Emv5Wmx---r1-33:-K3LFUvp_NwtIc7siuFK:-499146009) ([raw file](https://github.com/servo/servo/blob/c54b7b739739802ed51e7c15681dd208f2e4a9df/tests/wpt/grouping_formatter.py#L33)):_
Done.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

**Comment by jgraham** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[tests/wpt/grouping_formatter.py, line 53 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3LG-nwWltJkhSIGtpy:-K3LG-nwWltJkhSIGtpz:-1420639638) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L53)):_
Factor out the default values here and below

---

_[tests/wpt/grouping_formatter.py, line 58 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3LFriHpI8GF0MzMErA:-K3LFriHpI8GF0MzMErB:-746457823) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L58)):_
use `Exception as exception` (or I prefer `Exception as e`)

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

File: tests/wpt/grouping_formatter.py [hunk 2]
```diff
@@ -47,16 +46,25 @@ def __init__(self):
...
+            import blessings
+        except ImportError:
+            return u"\x1b[A", u"\x1b[K"
+
+        try:
+            self.terminal = blessings.Terminal()
+            return self.terminal.clear_eol, self.terminal.move_up
+        except Exception, exception:
+            sys.stderr.write("GroupingFormatter: Could not get terminal "
+                             "control characters: %s" % exception)
```
Comment (frewsxcv) [inline tests/wpt/grouping_formatter.py]: This probably needs a `\n` since `write` doesn't add one

**Comment by jgraham** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[tests/wpt/grouping_formatter.py, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3GRjeGJL9Emv5Wmx---r2-60:-K3LGthIglEL5sCCd4E4:-1755667099) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L60)):_
Actually I'm not sure this line doesn't do more harm than good since it causes a warning message that can't go through the mozlog infrastructure. I'd be inclined just to delete it.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

**Comment by mrobinson** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[tests/wpt/grouping_formatter.py, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3GRjeGJL9Emv5Wmx---r2-60:-K3LHLg7vsdgSdgToHU8:847023699) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L60)):_
I think it's still useful since it allows us to improve the script if there are problems finding the appropriate control codes. If it's deleted, I believe that the two try/except blocks should be merged. There isn't any point in isolated ImportError when we respond to all exceptions in the same way.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

**Comment by mrobinson** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[tests/wpt/grouping_formatter.py, line 53 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3LG-nwWltJkhSIGtpy:-K3LJBSPP9ziSgz0P-uS:-499146009) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L53)):_
Done.

---

_[tests/wpt/grouping_formatter.py, line 58 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3LFriHpI8GF0MzMErA:-K3LJBrMXso5dgPrGgjF:-499146009) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L58)):_
Done.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

**Comment by jgraham** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[tests/wpt/grouping_formatter.py, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3GRjeGJL9Emv5Wmx---r2-60:-K3LJbN5OMpPE4_XS48Q:-1238674272) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L60)):_
I guess it's probably OK like this, but it's not ideal.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

**Comment by mrobinson** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[tests/wpt/grouping_formatter.py, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8551#-K3GRjeGJL9Emv5Wmx---r2-60:-K3LLeblbh2gU9s8tMau:-1684008364) ([raw file](https://github.com/servo/servo/blob/2e04bd2f58305aa33c2b424e45afce50ff3157cb/tests/wpt/grouping_formatter.py#L60)):_
If you have a better approach for sending the error through machformatter, I'll be happy to implement it. I considered recording it and logging it after the logger was initialized, but that seemed a bit convoluted.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8551)_

File: tests/wpt/grouping_formatter.py [hunk 3]
```diff
@@ -47,16 +49,25 @@ def __init__(self):
...
 
+    def get_move_up_and_clear_eol_codes(self):
+        try:
+            import blessings
+        except ImportError:
+            return DEFAULT_MOVE_UP_CODE, DEFAULT_CLEAR_EOL_CODE
+
+        try:
+            self.terminal = blessings.Terminal()
+            return self.terminal.clear_eol, self.terminal.move_up
```
Comment (salty-horse) [inline tests/wpt/grouping_formatter.py]: clear_eol and move_up are in the wrong order :)

Comment (frewsxcv) [inline tests/wpt/grouping_formatter.py]: Excellent detective work @salty-horse :+1:

File: tests/wpt/grouping_formatter.py [hunk 4]
```diff
@@ -47,16 +49,25 @@ def __init__(self):
...
             return ""
-
-        # TODO(mrobinson, 8313): We need to add support for Windows terminals here.
-        erase_length = len(self.current_display)
-        if self.one_line:
-            return "\b \b" * erase_length
-        else:
-            return ("\033[F" + "\033[K") * len(self.current_display.splitlines())
+        return ((self.move_up + self.clear_eol) *
+                len(self.current_display.splitlines()))
```
Comment (salty-horse) [inline tests/wpt/grouping_formatter.py]: instead of ```len(s.splitlines())```, ```s.count('\n')``` should be equivalent (unless Windows does weird things, which I doubt)

File: tests/wpt/grouping_formatter.py (hunk 3 above)
Comment (mrobinson) [inline tests/wpt/grouping_formatter.py]: Nice catch!

Comment (frewsxcv) [inline tests/wpt/grouping_formatter.py]: @salty-horse You going to open a PR to fix it?

Comment (salty-horse) [inline tests/wpt/grouping_formatter.py]: Will do. Can I put my other suggestion in a separate commit in the same PR?

File: tests/wpt/grouping_formatter.py (hunk 4 above)
Comment (frewsxcv) [inline tests/wpt/grouping_formatter.py]: What advantages does your new solution have?

Comment (salty-horse) [inline tests/wpt/grouping_formatter.py]: It's slightly faster (did not measure). No need to create a list with new strings just to count its length. However:

1. It only works if you assume the last line ends with a newline (Simple fix if you want to support other cases, but I don't think it's an issue)
2. Only tested on Linux.

Comment (frewsxcv) [inline tests/wpt/grouping_formatter.py]: It might be better use `s.count(os.linesep)` for better Windows support, though I'm not too sure TBH

https://docs.python.org/2/library/os.html#os.linesep

Comment (mrobinson) [inline tests/wpt/grouping_formatter.py]: This is counting the number of lines in self.current_display, which is always assigned internally. I believe that means that Windpws line endings shouldn't pose any complications. As far as I can tell the .count("\n") approach should be fine.

Comment (salty-horse) [inline tests/wpt/grouping_formatter.py]: The rest of the strings are composed with \n, so I'm not sure what that will do :)
If no-one can confirm it's working on Windows/Mac, I'll just patch the other bit.

Comment (salty-horse) [inline tests/wpt/grouping_formatter.py]: I tested this on mac and windows, and it does what you expect. So seems fine to me:
```
'a\nb\nc\n'.count('\n')
```

Comment (jdm) [inline tests/wpt/grouping_formatter.py]: I would expect splitting a constant string on a constant separator to work the same on all platforms :) The real question is given multiline string content from an external source, does it contain `\r` that would cause difficulties when splitting on `\n`?

Comment (salty-horse) [inline tests/wpt/grouping_formatter.py]: You're right. Not sure why I tested that :P. In any case, there is no external source other than file names in that output.
